genai-perf/genai_perf/config/generate/optuna_objective_generator.py

```diff
@@ -0,0 +1,468 @@
+
+            if logger.level == logging.logging.DEBUG:
+                self._print_debug_score_info(trial_number, score)
+
+            if self._should_terminate_early(min_configs_to_search, trial_number):
+                logger.debug("Early termination threshold reached")
+                break
+            self._study.tell(trial, score)
+
+    ###########################################################################
+    # Measurement Methods
+    ###########################################################################
+    def set_measurement(self, measurement: Optional[RunConfigMeasurement]):
+        """
+        After profiling completes you must pass in the results
+        """
+        self._measurement_set = True
+        self._measurement = measurement
+
+    def _set_best_measurement(self, score: float = 0, trial_number: int = 0) -> None:
+        if not self._best_trial_number or score > self._best_score:
+            self._best_score = score
+            self._best_trial_number = trial_number
+
+    def _calculate_score(self) -> float:
+        if self._measurement:
+            score = self._measurement.get_score(self._measurement)
+        else:
+            score = OptunaObjectiveGeneratorDefaults.NO_MEASUREMENT_SCORE
+
+        return score
+
+    ###########################################################################
+    # General Search Space Methods
+    ###########################################################################
+    def _calculate_num_of_configs_in_search_space(self) -> int:
+        num_of_configs_in_search_space = 1
+        for model_name in self._config.model_names:
+            num_of_configs_in_search_space *= self._model_search_parameters[
+                model_name
+            ].number_of_total_possible_configurations()
+
+        return num_of_configs_in_search_space
+
+    ###########################################################################
+    # Minimum Search Space Methods
+    ###########################################################################
+    def _determine_minimum_number_of_configs_to_search(self) -> int:
+        min_trials_based_on_percentage_of_search_space = (
+            self._determine_trials_based_on_min_percentage_of_search_space()
+        )
+
+        min_configs_to_search = self._decide_min_between_percentage_and_trial_count(
+            min_trials_based_on_percentage_of_search_space
+        )
+
+        return min_configs_to_search
+
+    def _determine_trials_based_on_min_percentage_of_search_space(self) -> int:
+        total_num_of_possible_configs = self._calculate_num_of_configs_in_search_space()
+        min_trials_based_on_percentage_of_search_space = int(
+            total_num_of_possible_configs
+            * self._config.optimize.search_space_percentage.min
+            / 100
+        )
+
+        return min_trials_based_on_percentage_of_search_space
+
+    def _decide_min_between_percentage_and_trial_count(
```

Comment:
In a future refactor, I'd love to see the min/max functions refactored to be one, since they're mostly the same. There could be a flag to indicate whether to use the max or the min.

Reply:
I actually gave that some thought because it's almost a copy/paste, but there were enough differences (especially around all the debug) that I couldn't figure out a way to combine them cleanly.

Reply:
If I may give my own two cents, personally, I usually live by the principle that code needs to be refactored only after the third occurrence, not before. Having two almost identical copies of the same code is still acceptable in my book, it's when we reach three that we may want to think about it more.